### ISSUE TITLE
CURLE_WRITE_ERROR (23) as a retryable condition?

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nix#1311]
I just got this error:

```
Apr 07 17:15:24 my-ip.ec2.internal amazon-init-start[798]: curl: (23) Failed writing body (315 != 16366)
Apr 07 17:15:24 my-ip.ec2.internal amazon-init-start[798]: download of ‘https://cache.nixos.org/nar/1p3bi916yb3bn4h4jj3dv011zfga3b4wk41061gvdmpaiss39qia.nar.xz’ failed: No such file or directory
Apr 07 17:15:24 my-ip.ec2.internal amazon-init-start[798]: could not download ‘/nix/store/xjfycprfngknsdb8x4d2z88q99wpqkqw-cairo-1.14.8’ from any binary cache
```

It worked the second time around. Not really sure what it means, but do you think it makes sense to add 23 as a retryable condition? It seems like we can be a bit optimistic about what's retryable, since at the worst we'll retry things that keep failing a few times.

cc @edolstra

**Comment by copumpkin** [comment]
Here's another one that Nix doesn't consider retryable but that is trivially retriable:

```
Apr 08 12:46:48 my-ip.ec2.internal amazon-init-start[803]: error: unable to download ‘http://s3.amazonaws.com/mybucket/channel.tar.gz’: HTTP error 200 (curl error: Failure when receiving data from the peer)
```

Basically, I think S3 is fairly unfriendly HTTP-wise, but @edolstra would you object to me adding a few more retry conditions to enable this to work? As I said above, there doesn't seem to be too much downside to retrying more often than is technically sensible.

**Comment by edolstra** [comment]
@copumpkin We're already retrying on CURLE_RECV_ERROR...

**Comment by copumpkin** [comment]
@edolstra not sure what you're saying... are you talking about that second "error 200" message? That came straight out of `nixUnstable` and didn't appear to retry at all, based on my logs.

**Comment by edolstra** [comment]
@copumpkin You said we should retry "Failure when receiving data from the peer" but we're already doing that.

**Comment by copumpkin** [comment]
@edolstra all I can remark on is what I'm seeing from a real deployment of 1.12 (that log is from unstable). Maybe I should make that second one into a bug that it didn't actually seem to retry when the issue arose?

**Comment by copumpkin** [comment]
(for anyone following, it turns out that the "Failure when receiving data from the peer" is because the number of retries defaults to 1 and `nix-channel` doesn't change that)

**Comment by copumpkin** [comment]
This should be fixed by #1316 and #1317
